README.md
# NAME

Test::Class::Moose - Serious testing for serious Perl

# VERSION

version 0.99

# SYNOPSIS

    package TestsFor::DateTime;
    use Test::Class::Moose;
    use DateTime;

    # methods that begin with test_ are test methods.
    sub test_constructor {
        my $test = shift;
        $test->test_report->plan(3);    # strictly optional

        can_ok 'DateTime', 'new';
        my %args = (
            year  => 1967,
            month => 6,
            day   => 20,
        );
        isa_ok my $date = DateTime->new(%args), 'DateTime';
        is $date->year, $args{year}, '... and the year should be correct';
    }

    1;

# DESCRIPTION

See the [Test::Class::Moose home page](http://houseabsolute.github.io/test-class-moose/) for
a summary.

`Test::Class::Moose` is a powerful testing framework for Perl. Out of the box
you get:

- Reporting
- Extensibility
- Tagging tests
- Parallel testing
- Test inheritance
- Write your tests using Moose
- All the testing functions and behavior from Test::Most
- Event handlers for startup, setup, teardown, and shutdown of test classes

Better docs will come later. You should already know how to use Moose and
[Test::Class](https://metacpan.org/pod/Test%3A%3AClass).

# BASICS

## Inheriting from Test::Class::Moose

Just `use Test::Class::Moose`. That's all. You'll get all [Test::Most](https://metacpan.org/pod/Test%3A%3AMost) test
functions, too, along with `strict` and `warnings`. You can use all [Moose](https://metacpan.org/pod/Moose)
behavior, too.

When you `use Test::Class::Moose` it inserts itself as a parent class for
your test class. This means that if you try to use `extends` in your test
class you will break things unless you include `Test::Class::Moose` as a
parent. We recommend that you use roles in your test classes instead.

## Declare a test method

All method names that begin with `test_` are test methods. Methods that do
not are not test methods.

    sub test_this_is_a_method {
        my $test = shift;

        $test->this_is_not_a_test_method;
        ok 1, 'whee!';
    }

    sub this_is_not_a_test_method {
       my $test = shift;
       # but you can, of course, call it like normal
    }

You may specify `Test` and `Tests` method attributes, just like in
[Test::Class](https://metacpan.org/pod/Test%3A%3AClass) and the method will automatically be a test method, even if
does not start with `test_`:

    sub this_is_a_test : Test {
        pass 'we have a single test';
    }

    sub another_test_method : Tests { # like "no_plan"
        # a bunch of tests
    }

    sub yet_another_test_method : Tests(7) { # sets plan to 7 tests
        ...
    }

**Note**: Prior to version 0.51, this feature only worked if you had the
optional `Sub::Attribute` installed.

## Plans

No plans needed. The test suite declares a plan of the number of test classes.

Each test class is a subtest declaring a plan of the number of test methods.

Each test method relies on an implicit `done_testing` call.

If you prefer, you can declare a plan in a test method:

    sub test_something {
        my $test = shift;
        $test->test_report->plan($num_tests);
        ...
    }

Or with a `Tests` attribute:

    sub test_something : Tests(3) {
        my $test = shift;
        ...
    }

You may call `plan()` multiple times for a given test method. Each call to
`plan()` will add that number of tests to the plan.  For example, with a
method modifier:

    before 'test_something' => sub {
        my $test = shift;
        $test->test_report->plan($num_extra_tests);

        # more tests
    };

Please note that if you call `plan`, the plan will still show up at the end
of the subtest run, but you'll get the desired failure if the number of tests
run does not match the plan.

## Inheriting from another Test::Class::Moose class

List it as the `extends` in the import list. If the base class does not use
(or extend) Test::Class::Moose, then a compile-time error is thrown.

    package TestsFor::Some::Class::Subclass;
    use Test::Class::Moose extends => 'TestsFor::Some::Class';

    sub test_me {
        my $test  = shift;
        my $class = $test->test_class;
        ok 1, "I overrode my parent! ($class)";
    }

    before 'test_this_baby' => sub {
        my $test  = shift;
        my $class = $test->test_class;
        pass "This should run before my parent method ($class)";
    };

    sub this_should_not_run {
        my $test = shift;
        fail "We should never see this test";
    }

    sub test_this_should_be_run {
        for ( 1 .. 5 ) {
            pass "This is test number $_ in this method";
        }
    }

    1;

## Skipping Test::Most

By default, when you `use Test::Class::Moose` in your own test class, it
exports all the subs from [Test::Most](https://metacpan.org/pod/Test%3A%3AMost) into your class. If you'd prefer to
import a different set of test tools, you can pass `bare => 1` when using
`Test::Class::Moose`:

    use Test::Class::Moose bare => 1;

When you pass this, `Test::Class::Moose` will not export [Test::Most](https://metacpan.org/pod/Test%3A%3AMost)'s subs
into your class. You will have to explicitly import something like
[Test::More](https://metacpan.org/pod/Test%3A%3AMore) or [Test2::Tools::Compare](https://metacpan.org/pod/Test2%3A%3ATools%3A%3ACompare) in order to actually perform tests.

## Custom Test Toolkits

If you'd like to provide a custom set of test modules to all of your test
classes, this is easily done with [Import::Into](https://metacpan.org/pod/Import%3A%3AInto):

    package MM::Test::Class::Moose;

    use strict;
    use warnings;
    use namespace::autoclean ();

    use Import::Into;
    use Test::Class::Moose ();
    use Test::Fatal;
    use Test::More;

    sub import {
        my @imports = qw(
          Test::Class::Moose
          namespace::autoclean
          Test::Fatal
          Test::More
        );

        my $caller_level = 1;
        $_->import::into($caller_level) for @imports;
    }

You could also create a kit in a separate module like `My::Test::Kit` using
[Test::Kit](https://metacpan.org/pod/Test%3A%3AKit) and then simply export that from your `My::Test::Class::Moose`
module with [Import::Into](https://metacpan.org/pod/Import%3A%3AInto).

# TEST CONTROL METHODS

The test control methods are there to ensure that your tests are running with
all of the resources they need. For example, database transactions might be
started before a test method and rolled back after the test method. Fixtures
needed for every test might be loaded and cleaned up. However you use them,
it's important to understand when and how they're run.

- 1. `test_startup` — Runs once when the test class starts up
- 2. `test_setup` - Runs before each test method
- 3. `test_teardown` - Runs after each test method
- 4. `test_shutdown` - Runs once when the test class shuts down

**Important**: Do not run tests in test control methods. This will cause the
test control method to fail (this is a feature, not a bug).  If a test control
method fails, the class/method will fail and testing for that class should
stop.

The available test control methods are:

## `test_startup`

    sub test_startup {
       my $test = shift;
       $test->next::method;   # run this before your test_startup code
       # more startup
    }

Runs at the start of each test class. Quite often the base class that you
inherit from will have its own `test_startup` code running (such as starting
a database transaction or connecting to an external resource). You almost
always want to call `$test->next::method` _before_ your own setup code.
This ensures that the environment is set up to safely run your code. For
example, if the parent `test_startup` starts a database transaction with the
expectation that the `test_teardown` will end the database transactions, you
can safely load database fixtures _after_ that is run.

If you need to know the name of the class you're running this in (though
usually you shouldn't), use `$test->test_class`, or you can do this:

    sub test_startup {
        my $test                 = shift;
        my $report               = $test->test_report;
        my $instance             = $report->current_instance->name;
        my $upcoming_test_method = $report->current_method->name;
        ...
    }

The `$test->test_report` object is a [Test::Class::Moose::Report::Instance](https://metacpan.org/pod/Test%3A%3AClass%3A%3AMoose%3A%3AReport%3A%3AInstance)
object.

## `test_setup`

    sub test_setup {
       my $test = shift;
       $test->next::method;    # run this before your test_setup code
       # more setup
    }

Runs at the start of each test method. If you must know the name of the test
you're about to run, you can do this:

    sub test_setup {
       my $test = shift;
       $test->next::method;
       my $test_method = $test->test_report->current_method->name;
       # do something with it
    }

## `test_teardown`

    sub test_teardown {
       my $test = shift;
       # more teardown
       $test->next::method;    # run this after your test_teardown code
    }

Runs at the end of each test method.

By default, this is not run if the test class is skipped entirely. You can
override the `run_control_methods_on_skip` in your class to return a true
value in order to force this method to be run when the class is skipped.

## `test_shutdown`

    sub test_shutdown {
        my $test = shift;
        # more teardown
        $test->next::method;    # run this after your test_shutdown code
    }

Runs at the end of each test class.

By default, this is not run if the test class is skipped entirely. You can
override the `run_control_methods_on_skip` in your class to return a true
value in order to force this method to be run when the class is skipped.

## Overriding Test Control Methods

To override a test control method, just remember that this is OO:

    sub test_setup {
        my $test = shift;
        $test->next::method; # call parent test_setup
        # more setup code here
    }

# TEST CLASS INSTANCES

**This feature is still considered experimental.**

By default, each test class you create will be instantiated once. However, you
can tell the [Test::Class::Moose::Runner](https://metacpan.org/pod/Test%3A%3AClass%3A%3AMoose%3A%3ARunner) to create multiple instances of a
test class.

To do this, simply consume the
[Test::Class::Moose::Role::ParameterizedInstances](https://metacpan.org/pod/Test%3A%3AClass%3A%3AMoose%3A%3ARole%3A%3AParameterizedInstances) role in your test
class. This role requires you to implement a `_constructor_parameter_sets`
method in your test class. That method will be called as a _class method_. It
is expected to return a list of key/value pairs. The keys are the name of the
instance and the values are hashrefs of attributes to be passed to your test
class's constructor. Here's a really dumb example:

    package TestsFor::PlainAndFancy;
    use Test::Class::Moose;
    with 'Test::Class::Moose::Role::ParameterizedInstances';

    has is_fancy => (
        is       => 'ro',
        isa      => 'Bool',
        required => 1,
    );

    sub _constructor_parameter_sets {
        my $class = shift;
        return (
            "$class - plain" => { is_fancy => 0 },
            "$class - fancy" => { is_fancy => 1 },
        );
    }

    sub test_something { ... }

The test runner will run all the test methods in your class _once per
instance_, and each instance will be run in its own subtest. You can
dynamically decide to skip your test class completely by having
`_constructor_parameter_sets` return an empty list.

Note that this feature has great potential for abuse, so use it
cautiously. That said, there are cases where this feature can greatly simplify
your test code.

# RUNNING THE TEST SUITE

See the docs for [Test::Class::Moose::Runner](https://metacpan.org/pod/Test%3A%3AClass%3A%3AMoose%3A%3ARunner) for details on running your
test suite. If you'd like to get up and running quickly, here's a very simple
test file you can use:

    use Test::Class::Moose::CLI;
    Test::Class::Moose::CLI->new_with_options->run;

Put this in a file like `t/run-test-class.t`. When you run it with prove it
will load all the test classes defined in `t/lib` and run them sequentially.

See the documentation for [Test::Class::Moose::CLI](https://metacpan.org/pod/Test%3A%3AClass%3A%3AMoose%3A%3ACLI) on the options you can
pass when running tests.

## Skipping Classes and Methods

If you wish to skip a class, set the reason in the `test_startup` method.

    sub test_startup {
        my $test = shift;
        $test->test_skip("I don't want to run this class");
    }

If you are using [test class instances](#test-class-instances), you
can also make `_constructor_parameter_sets` return an empty list,
which will result in the class being skipped.

Note that if you run `test_skip`, the `test_shutdown` method will also be
skipped. This is due to the assumption that you might not have run any setup
code and thus you don't need shutdown code. However, if you do need to run
shutdown, you can override the `run_control_methods_on_skip` method to return
true:

    sub run_control_methods_on_skip {1}

If you wish to skip an individual method, do so in the `test_setup` method.

    sub test_setup {
        my $test = shift;
        my $test_method = $test->test_report->current_method;

        if ( 'test_time_travel' eq $test_method->name ) {
            $test->test_skip("Time travel not yet available");
        }
    }

## The "Tests" and "Test" Attributes

If you're comfortable with [Test::Class](https://metacpan.org/pod/Test%3A%3AClass), you know that test methods methods are
declared in [Test::Class](https://metacpan.org/pod/Test%3A%3AClass) with `Test` (for a method with a single test) or
`Tests`, for a method with multiple tests. This also works for
`Test::Class::Moose`. Test methods declared this way do not need to start
with `test_`.

    sub something_we_want_to_check : Test {
        # this method may have only one test
    }

    sub something_else_to_check : Tests {
        # this method may have multiple tests
    }

    sub another_test_method : Tests(3) {
        # this method must have exactly 3 tests
    }

If a test method overrides a parent test method and calls it, their plans will
be added together:

    package TestsFor::Parent;

    use Test::Class::Moose;

    sub some_test : Tests(3) {
        # three tests
    }

And later:

    package TestsFor::Child;

    use Test::Class::Moose extends => 'TestsFor::Parent';

    sub some_test : Tests(2) {
        my $test = shift;
        $test->next::method;
        # 2 tests here
    }

In the above example, `TestsFor::Parent::some_test` will run three tests, but
`TestsFor::Child::some_test` will run _five_ tests (two tests, plus the
three from the parent).

Note that if a plan is explicitly declared, any modifiers or overriding
methods calling the original method will also have to assert the number of
tests to ensure the plan is correct. The above `TestsFor::Parent` and
`TestsFor::Child` code would fail if the child's `some_test` method
attribute was `Tests` without the number of tests asserted.

Do not use `Test` or `Tests` with test control methods because you don't run
tests in those.

## Tagging Methods

Sometimes you want to be able to assign metadata to help you better manage
your test suite. You can do this with tags:

    sub test_save_poll_data : Tags(api network) {
        ...
    }

Tags are strictly optional and you can provide one or more tags for each test
method with a space separated list of tags. You can use this to filter your
tests suite, if desired. For example, if your network goes down and all tests
which rely on a network are tagged with `network`, you can skip those tests
with this:

    Test::Class::Moose::Runner->new( exclude_tags => 'network' )->runtests;

Or maybe you want to run all `api` and `database` tests, but skip those
marked `deprecated`:

    Test::Class::Moose::Runner->new(
        include_tags => [qw/api database/],
        exclude_tags => 'deprecated',
    )->runtests;

You can also inspect tags within your test classes:

    sub test_setup {
        my $test          = shift;
        my $method_to_run = $test->test_report->current_method;
        if ( $method_to_run->has_tag('db') ) {
            $test->load_database_fixtures;
        }
    }

Tagging support relies on [Sub::Attribute](https://metacpan.org/pod/Sub%3A%3AAttribute). If this module is not available,
`include_tags` and `exclude_tags` will be ignored, but a warning will be
issued if those are seen. Prior to version 0.51, `Sub::Attribute` was
optional. Now it's mandatory, so those features should always work.

# THINGS YOU CAN OVERRIDE

... but probably shouldn't.

As a general rule, methods beginning with `/^test_/` are reserved for
[Test::Class::Moose](https://metacpan.org/pod/Test%3A%3AClass%3A%3AMoose). This makes it easier to remember what you can and
cannot override. However, any test with `Test` or `Tests` are test methods
regardless of their names.

## `test_report`

    my $report = $test->test_report;

Returns the [Test::Class::Moose::Report](https://metacpan.org/pod/Test%3A%3AClass%3A%3AMoose%3A%3AReport) object. Useful if you want to do
your own reporting and not rely on the default output provided with the
`statistics` boolean option.

You can also call it in test classes (most useful in the `test_setup()` method):

    sub test_setup {
        my $test = shift;
        $self->next::method;
        my $report = $test->test_report;
        my $instance = $test->current_instance;
        my $method = $test->current_method; # the test method we're about to run
        if ( $method->name =~ /customer/ ) {
            $test->load_customer_fixture;
        }
        # or better still
        if ( $method->has_tag('customer') ) {
            $test->load_customer_fixture;
        }
    }

## `test_class`

    my $class = $test->test_class;

Returns the name for this test class. Useful if you rebless an object (such as
applying a role at runtime) and don't want to lose the original class name.

## `test_methods`

You may override this in a subclass. Currently returns all methods in a test
class that start with `test_` (except for the test control methods).

Please note that the behavior for `include` and `exclude` is also contained
in this method. If you override it, you will need to account for those
yourself.

## `import`

Sadly, we have an `import` method. This is used to automatically provide you
with all of the [Test::Most](https://metacpan.org/pod/Test%3A%3AMost) behavior.

# SAMPLE TAP OUTPUT

We use nested tests (subtests) at each level:

    1..2
    #
    # Executing tests for TestsFor::Basic::Subclass
    #
        1..3
        # TestsFor::Basic::Subclass->test_me()
            ok 1 - I overrode my parent! (TestsFor::Basic::Subclass)
            1..1
        ok 1 - test_me
        # TestsFor::Basic::Subclass->test_this_baby()
            ok 1 - This should run before my parent method (TestsFor::Basic::Subclass)
            ok 2 - whee! (TestsFor::Basic::Subclass)
            1..2
        ok 2 - test_this_baby
        # TestsFor::Basic::Subclass->test_this_should_be_run()
            ok 1 - This is test number 1 in this method
            ok 2 - This is test number 2 in this method
            ok 3 - This is test number 3 in this method
            ok 4 - This is test number 4 in this method
            ok 5 - This is test number 5 in this method
            1..5
        ok 3 - test_this_should_be_run
    ok 1 - TestsFor::Basic::Subclass
    #
    # Executing tests for TestsFor::Basic
    #
        1..2
        # TestsFor::Basic->test_me()
            ok 1 - test_me() ran (TestsFor::Basic)
            ok 2 - this is another test (TestsFor::Basic)
            1..2
        ok 1 - test_me
        # TestsFor::Basic->test_this_baby()
            ok 1 - whee! (TestsFor::Basic)
            1..1
        ok 2 - test_this_baby
    ok 2 - TestsFor::Basic
    # Test classes:    2
    # Test methods:    5
    # Total tests run: 11
    ok
    All tests successful.
    Files=1, Tests=2,  2 wallclock secs ( 0.03 usr  0.00 sys +  0.27 cusr  0.01 csys =  0.31 CPU)
    Result: PASS

# REPORTING

See [Test::Class::Moose::Report](https://metacpan.org/pod/Test%3A%3AClass%3A%3AMoose%3A%3AReport) for more detailed information on reporting.

Reporting features are subject to change.

Sometimes you want more information about your test classes, it's time to do
some reporting. Maybe you even want some tests for your reporting. If you do
that, run the test suite in a subtest (because the plans will otherwise be
wrong).

    #!/usr/bin/env perl
    use lib 'lib';
    use Test::Most;
    use Test::Class::Moose::Load qw(t/lib);
    use Test::Class::Moose::Runner;

    my $test_suite = Test::Class::Moose::Runner->new;

    subtest 'run the test suite' => sub {
        $test_suite->runtests;
    };
    my $report = $test_suite->test_report;

    foreach my $class ( $report->all_test_instances ) {
        my $class_name = $class->name;
        ok !$class->is_skipped, "$class_name was not skipped";
        ok $class->passed, "$class_name passed";

        subtest "$class_name methods" => sub {
            foreach my $method ( $class->all_test_methods ) {
                my $method_name = $method->name;
                ok $method->passed, "$method_name passed";

                ok !$method->is_skipped, "$method_name was not skipped";
                cmp_ok $method->num_tests, '>', 0,
                  '... and some tests should have been run';
                diag "Run time for $method_name: ".$method->time->duration;
            }
        };
        my $time   = $class->time;
        diag "Run time for $class_name: ".$class->time->duration;

        my $real   = $time->real;
        my $user   = $time->user;
        my $system = $time->system;
        # do with these as you will
    }
    diag "Number of test classes: "   . $report->num_test_classes;
    diag "Number of test instances: " . $report->num_test_instances;
    diag "Number of test methods: "   . $report->num_test_methods;
    diag "Number of tests:        "   . $report->num_tests;

    done_testing;

If you just want to output reporting information, you do not need to run the
test suite in a subtest:

    my $test_suite = Test::Class::Moose::Runner->new->runtests;
    my $report     = $test_suite->test_report;
    ...

Or even shorter:

    my $report = Test::Class::Moose::Runner->new->runtests->test_report;

# EXTRAS

If you would like [Test::Class::Moose](https://metacpan.org/pod/Test%3A%3AClass%3A%3AMoose) to take care of loading your classes
for you, see [Test::Class::Moose::Role::AutoUse](https://metacpan.org/pod/Test%3A%3AClass%3A%3AMoose%3A%3ARole%3A%3AAutoUse) in this distribution.

# DEPRECATIONS AND BACKWARDS INCOMPATIBILITIES

## Version 0.79

- The [Test::Class::Moose::Config](https://metacpan.org/pod/Test%3A%3AClass%3A%3AMoose%3A%3AConfig) class's `args` method is now
deprecated. This was a holdover from when Test::Class::Moose was both a parent
class for your test classes and the test class runner.

## Version 0.77

- The passing of the report object as an argument to test methods and test
control methods is now deprecated. You can get the report from the test class
object itself via the `$test->test_report` method.
- The `Test::Class::Moose->runtests` method has been removed. Use
[Test::Class::Moose::Runner](https://metacpan.org/pod/Test%3A%3AClass%3A%3AMoose%3A%3ARunner) to run your test classes.
- The `Test::Class::Moose::Role::Parallel` role has been removed. This has not
done anything except issue a warning since version 0.55.

## Version 0.75

- The `test_teardown method` is no longer run when a test is skipped unless
`run_control_methods_on_skip` returns a true value. The `test_teardown`
method was never intended to be run unconditionally.
- Parallel testing now parallelizes test classes rather than individual test
instances. This is only relevant if your test suite contains parameterized
test classes. This is slightly less efficient, but made the internal test
running code much simpler and made it possible to fix reporting for parallel
test runs.
- The [Test::Class::Moose::Config](https://metacpan.org/pod/Test%3A%3AClass%3A%3AMoose%3A%3AConfig) `builder` method has been removed.
- The [Test::Class::Moose::Runner](https://metacpan.org/pod/Test%3A%3AClass%3A%3AMoose%3A%3ARunner) `builder` method has been removed.

## Version 0.67

- The [Test::Class::Moose::Report](https://metacpan.org/pod/Test%3A%3AClass%3A%3AMoose%3A%3AReport) class's `all_test_classes` method is un-deprecated

    This method now returns a list of [Test::Class::Moose::Report::Class](https://metacpan.org/pod/Test%3A%3AClass%3A%3AMoose%3A%3AReport%3A%3AClass)
    objects. A class report contains one or more instance reports.

- Removed the [Test::Class::Moose::Report::Instance](https://metacpan.org/pod/Test%3A%3AClass%3A%3AMoose%3A%3AReport%3A%3AInstance)'s error
attribute. Contrary to the documentation, this attribute was never populated.
- Renamed the [Test::Class::Moose::Report::Method](https://metacpan.org/pod/Test%3A%3AClass%3A%3AMoose%3A%3AReport%3A%3AMethod) `instance_report` method to
`instance`. This is a better match for other report-related methods, which
don't include a "\_report" suffix.
- Removed the long-deprecated `tests_run` methods from
[Test::Class::Moose::Report](https://metacpan.org/pod/Test%3A%3AClass%3A%3AMoose%3A%3AReport) and [Test::Class::Moose::Report::Method](https://metacpan.org/pod/Test%3A%3AClass%3A%3AMoose%3A%3AReport%3A%3AMethod).
- Removed the long-deprecated `Test::Class::Moose::Report::Method->add_to_plan method`.

## Version 0.55

- Running tests with Test::Class::Moose is deprecated - use [Test::Class::Moose::Runner](https://metacpan.org/pod/Test%3A%3AClass%3A%3AMoose%3A%3ARunner)

    As of version 0.55, running tests and being a test class have been
    separated. Your test classes should continue to `use Test::Class::Moose`, but
    your test runner script should use [Test::Class::Moose::Runner](https://metacpan.org/pod/Test%3A%3AClass%3A%3AMoose%3A%3ARunner):

        use Test::Class::Moose::Load 't/lib';
        use Test::Class::Moose::Runner;

Changes
Revision history for Perl distribution Test-Class-Moose

{{$NEXT}}

0.99     2021-02-07

  [ENHANCEMENTS]

  * Greatly expanded documentation for the test control methods.

  [BUG FIXES]

  * Fixed a bug in the method report's `has_tag` method. This would die no
    matter how it was called. Fixed by Iain Loasby. Fixes GH #109. GH #110.


0.98     2019-10-26

  [BUG FIXES]

  * Escape the directory name when using it in regex while loading files in
    Test::Class::Moose::Load. Paths with special characters were not loading
    modules properly. Fixed by jrubinator. GH #101.


0.97     2019-04-21

  [BUG FIXES]

  * Require Test2 1.302118, not 1.1302118 (note the extra 1 after the
    decimal). Fixed by Christopher Layne. GH #99.


0.96     2018-12-21

  [ENHANCEMENTS]

  * Converted uses of JSON.pm to JSON::MaybeXS. Implemented by Karen
    Etheridge. PR #98.


0.95     2018-11-03

  [ENHANCEMENTS]

  * The CLI class/role now accepts a --test-lib-dirs option. This lets you
    tell it what directories contain test classes from the CLI.


0.94     2018-08-18

  [BUG FIXES]

  * Replace all cases of "use 5.10.0" with "use 5.010000". The former was
    causing some versions of Perl to emit warnings that cause tests to fail.


0.93     2018-08-14

  [ENHANCEMENTS]

  * The CLI class (and role) now allow you to pass a directory for the
  --classes option. This directory is searched recursively for classes to be
  loaded. Based on a feature request from jrubinator. GH #91.

  [BUG FIXES]

  * Extending a TCM class with a non-existing or non-TCM class now delivers a
    fatal error instead of silently ignoring the test class. Implemented by
    Harald Jörg. PR #93, fixes GH #70.

0.92     2017-12-23

  [ENHANCEMENTS]

  * Added a new "executor_roles" parameter to the Test::Class::Moose::Runner
    constructor. This allows you to apply roles to the executor (the object
    which executes test classes) as a way to extend the default TCM
    behavior. See the documentation for Test::Class::Moose::Role::Executor for
    methods which you can wrap.

  [BACKWARDS INCOMPATIBILITIES]

  * Removed the never-documented Test::Class::Moose::Report's is_parallel
    attribute. This was present because in older versions of TCM, reporting
    was broken when tests were run in parallel.

0.91     2017-12-07

  [BUG FIXES]

  * Use sum0 instead of sum in the num_test_instances() method for the report
    object. Fixed by Larry Leszczynski. GH #87.

0.90     2017-11-28

  [BUG FIXES]

  * Require Test2 1.1302118 per Chad Granum's recommendation. This version
    includes some fixes for bugs that could cause tests to hang.

0.89     2017-11-19

  [ENHANCEMENTS]

  * Moved the check for bad class names in Test::Class::Moose::Role::Executor
    to its own method to facilitate overriding just that portion of the
    role. Based on PR #85 from Chuck Adams, and addresses #84.

0.88     2017-11-04

  [BUG FIXES]

  - When no tests were run, the num_tests_run() method for the report object
    would return undef instead of 0. Reported by Johan Lindstrom. GH #83.

0.87     2017-09-12

  [BUG FIXES]

  * Fixed link to the TCM home page in the docs. Fixed by Paul Williams. GH
    #82.

0.86     2017-09-05

  [BUG FIXES]

  * The change to add start_time and end_time methods to all reports
    completely broke timing data, so all tests reported taking essentially
    zero time. Reported by Andy Armstrong. GH #81.

0.85     2017-07-29

  [ENHANCEMENTS]

  * All report objects now have start_time and end_time methods.

  * Added support for the "bare => 1" import option to
    Test::Class::Moose::Role. Implemented by Desmond Daignault. PR #80.

0.84     2017-05-15

  [ENHANCEMENTS]

  * Make it possible to disable the progress output shown when running tests
    in parallel. This does not play nicely with things like
    TAP::Harness::Archive and prove's --merge flag. The output ends up
    interleaved with test output, breaking the TAP.

0.83     2017-05-10

  [ENHANCEMENTS]

  * Test::Most is only loaded if you do not pass "bare => 1" when importing
    Test::Class::Moose.

  [BUG FIXES]

  * The CLI implementation released in 0.82 was broken in regards to the
    --classes parameter. This would not actually load the specified classes at
    all, because the test lib dirs were not being added to @INC.

0.82     2017-04-05

  [ENHANCEMENTS]

  * Added a Test::Class::Moose::CLI class. This provides a ready-to-go
    interface for handling CLI parameters passed to the test file that runs
    your Test::Class::Moose tests. This lets you do things like run "prove -lv
    t/tcm.t :: --classes TestFor::My::Thing" without having to write your own
    code to handle this.

    This new CLI class is entirely implemented by a role,
    Test::Class::Moose::Role::CLI, that has well-documented hooks for you to
    write your own CLI class on top of. See the docs for the class and the
    role for more details.

0.81     2017-03-16

  [ENHANCEMENTS]

  * If you pass one or more test class names to the Runner class explicitly
    and those classes are either not loaded or not subclasses of
    Test::Class::Moose, you'll get a better error message. Previously this
    would die with an unhelpful error about missing methods. Reported by Andy
    Cunningham with additional clarification from hoppfrosch. GH #57.

0.80     2016-09-25

  [BUG FIXES]

  * Depend on Test2 1.302059. This version fixes an issue where subtest output
    from test methods was not showing up. Reported by Ilya Chesnokov. GH #67.

0.79     2016-09-25

  [BACKWARDS INCOMPATIBILITIES]

  * The value of Test::Class::Moose::Config->args is no longer passed to the
    constructor for each test class instance. This was a holdover from when
    Test::Class::Moose was both a parent class for your test classes and the
    test class runner. Fixes GH #68. Reported by Noel Maddy.

0.78     2016-08-21

  [ENHANCEMENTS]

  * Updated the included Test::Class::Moose::Tutorial docs.


Makefile.PL
# This Makefile.PL for Test-Class-Moose was generated by
# Dist::Zilla::Plugin::DROLSKY::MakeMaker 1.18
# and Dist::Zilla::Plugin::MakeMaker::Awesome 0.48.
# Don't edit it but the dist.ini and plugins used to construct it.

use strict;
use warnings;

use 5.010000;
use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Serious testing for serious Perl",
  "AUTHOR" => "Curtis \"Ovid\" Poe <ovid\@cpan.org>, Dave Rolsky <autarch\@urth.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Test-Class-Moose",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010000",
  "NAME" => "Test::Class::Moose",
  "PREREQ_PM" => {
    "Benchmark" => 0,
    "Carp" => 0,
    "Class::MOP" => 0,
    "Exporter" => 0,
    "File::Find" => 0,
    "File::Spec" => 0,
    "Import::Into" => 0,
    "JSON::MaybeXS" => 0,
    "List::SomeUtils" => 0,
    "List::Util" => 0,
    "Module::Runtime" => 0,
    "Module::Util" => 0,
    "Moose" => "2.0000",
    "Moose::Role" => "2.0000",
    "Moose::Util" => 0,
    "Moose::Util::TypeConstraints" => 0,
    "MooseX::Getopt" => "0.71",
    "MooseX::Getopt::Dashes" => 0,
    "Package::DeprecationManager" => "0.16",
    "Parallel::ForkManager" => 0,
    "Scalar::Util" => 0,
    "Sub::Attribute" => 0,
    "TAP::Formatter::Color" => "3.29",
    "Test2" => "1.302118",
    "Test2::API" => 0,
    "Test2::AsyncSubtest" => "0.000018",
    "Test2::IPC" => 0,
    "Test2::Tools::AsyncSubtest" => 0,
    "Test::Most" => 0,
    "Try::Tiny" => 0,
    "namespace::autoclean" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "File::Temp" => 0,
    "FindBin" => 0,
    "Test2::Tools::Basic" => 0,
    "Test2::Tools::Class" => 0,
    "Test2::Tools::Compare" => 0,
    "Test2::Tools::Subtest" => 0,
    "Test2::V0" => 0,
    "Test::Builder" => 0,
    "Test::More" => "1.302015",
    "Test::Requires" => 0,
    "Test::Warnings" => 0,
    "lib" => 0
  },
  "VERSION" => "1.00",
  "test" => {
    "TESTS" => "t/*.t"
  }
);

my %FallbackPrereqs = (
  "Benchmark" => 0,
  "Carp" => 0,
  "Class::MOP" => 0,
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Find" => 0,
  "File::Spec" => 0,
  "File::Temp" => 0,
  "FindBin" => 0,
  "Import::Into" => 0,
  "JSON::MaybeXS" => 0,
  "List::SomeUtils" => 0,
  "List::Util" => 0,
  "Module::Runtime" => 0,
  "Module::Util" => 0,
  "Moose" => "2.0000",
  "Moose::Role" => "2.0000",
  "Moose::Util" => 0,
  "Moose::Util::TypeConstraints" => 0,
  "MooseX::Getopt" => "0.71",
  "MooseX::Getopt::Dashes" => 0,
  "Package::DeprecationManager" => "0.16",
  "Parallel::ForkManager" => 0,
  "Scalar::Util" => 0,
  "Sub::Attribute" => 0,
  "TAP::Formatter::Color" => "3.29",
  "Test2" => "1.302118",
  "Test2::API" => 0,
  "Test2::AsyncSubtest" => "0.000018",
  "Test2::IPC" => 0,
  "Test2::Tools::AsyncSubtest" => 0,
  "Test2::Tools::Basic" => 0,
  "Test2::Tools::Class" => 0,
  "Test2::Tools::Compare" => 0,
  "Test2::Tools::Subtest" => 0,
  "Test2::V0" => 0,
  "Test::Builder" => 0,
  "Test::More" => "1.302015",
  "Test::Most" => 0,
  "Test::Requires" => 0,
  "Test::Warnings" => 0,
  "Try::Tiny" => 0,
  "lib" => 0,
  "namespace::autoclean" => 0,
  "strict" => 0,
  "warnings" => 0
);

unless ( eval { ExtUtils::MakeMaker->VERSION('6.63_03') } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
